### PR TITLE
golang: verify handle path trailing /

### PR DESCRIPTION
**Pull request by tl-marco-tormento**
https://github.com/TrueLayer/truelayer-signing/issues/80 (option 2) for golang.

### CODE REVIEW

File: go/verify/verifier.go [hunk 1]
```diff
@@ -154,17 +154,32 @@ func (v *Verifier) Verify(tlSignature string) error {
...
 }
 
+func (v *Verifier) handleTrailingSlashRetry() (string, bool) {
+	if strings.HasSuffix(v.path, "/") {
+		return strings.TrimSuffix(v.path, "/"), false
+	} else {
+		return v.path, true
+	}
+}
+
```
Comment (luk4ward) [inline go/verify/verifier.go:182]:
```suggestion
func (v *Verifier) handleTrailingSlashRetry() (string, bool) {
	return strings.TrimSuffix(v.path, "/"), strings.HasSuffix(v.path, "/")
}

```

File: go/sign/signer.go [hunk 2]
```diff
@@ -125,11 +125,14 @@ func (s *Signer) Sign() (string, error) {
...
 
 // BuildV2SigningPayload builds a v2 signing payload.
-func BuildV2SigningPayload(method string, path string, headers *orderedmap.OrderedMap, body []byte) []byte {
+func BuildV2SigningPayload(method string, path string, headers *orderedmap.OrderedMap, body []byte, addPathTrailingSlash bool) []byte {
 	payload := make([]byte, 0)
 	payload = append(payload, []byte(strings.ToUpper(method))...)
 	payload = append(payload, []byte(" ")...)
 	payload = append(payload, []byte(path)...)
+	if addPathTrailingSlash {
+		payload = append(payload, []byte("/")...)
```
Comment (conradludgate) [inline go/sign/signer.go:134]:
```suggestion
		payload = append(payload, '/')
```

File: go/tlsigning_test.go [hunk 3]
```diff
@@ -477,3 +477,51 @@ func TestSetJku(t *testing.T) {
 	assert.Nilf(err, "jws header extraction failed: %v", err)
 	assert.Equal("https://webhooks.truelayer.com/.well-known/jwks", jwsHeader.Jku)
 }
+
+// Signing a path with a single trailing slash & trying to verify
+// without that slash should still work. See #80.
+func VerifyWithoutSignedTrailingSlash(t *testing.T) {
```
Comment (conradludgate) [inline go/tlsigning_test.go:483]: Doesn't this need a Test prefix name?

**Review by conradludgate — COMMENTED** [review]
I think the tests need fixing

File: go/tlsigning_test.go (hunk 3 above)
Comment (tl-marco-tormento) [reply go/tlsigning_test.go]: Definitely... 🤦🏻

File: go/verify/verifier.go [hunk 4]
```diff
@@ -154,17 +154,32 @@ func (v *Verifier) Verify(tlSignature string) error {
...
+	payload := fmt.Sprintf("%s.%s", tlSignatureData.HeaderBase64, base64.RawURLEncoding.EncodeToString(signingPayload))
 	err = crypto.VerifyES512(publicKey, []byte(payload), tlSignatureData.Signature)
 	if err != nil {
-		return errors.NewJwsError(fmt.Sprintf("signature verification failed: %v", err))
+		// try again with/without a trailing slash (#80)
+		newPath, addPathTrailingSlash := v.handleTrailingSlashRetry()
+		signingPayload := sign.BuildV2SigningPayload(v.method, newPath, orderedHeaders, v.body, addPathTrailingSlash)
+		payload := fmt.Sprintf("%s.%s", tlSignatureData.HeaderBase64, base64.RawURLEncoding.EncodeToString(signingPayload))
+		retryErr := crypto.VerifyES512(publicKey, []byte(payload), tlSignatureData.Signature)
+		if retryErr != nil {
```
Comment (luk4ward) [inline go/verify/verifier.go:166]:
```suggestion
		if retryErr := crypto.VerifyES512(publicKey, []byte(payload), tlSignatureData.Signature); retryErr != nil {
```

File: go/verify/verifier.go (hunk 1 above)
Comment (tl-marco-tormento) [reply go/verify/verifier.go:182]: I don't think these are equivalent.
Maybe we should negate the second return value?

Comment (luk4ward) [reply go/verify/verifier.go]: oh yeah I missed the `!` 🤦

**Review by luk4ward — APPROVED** [review]

**Review by conradludgate — APPROVED** [review]

**Pull request merged**